Correctly clear lines for multi-line progress bar messages

grasmash opened the pull request:
Measure the length of individual lines when calculating progress bar message length.

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no
| License       | MIT

When a multiline message is added to a progress bar, the progress bar measures the length of all lines combined in a single string without considering whether that string contains new line characters. It is not accurately measuring the number of lines that are displayed on the terminal. 

On the terminal, the actual number of new lines is dictated by a combination of new line characters and text extending beyond the terminal width. Due to this bug, the progress bar sometimes clears too many lines when the progress bar is advanced.

The progress bar should instead consider the overlap of $this->formatLineCount and the actual number of lines in the message and then add any remainder.

chalasr commented:
Thanks for this. Should it be applied to 4.4 as well?
Also, can you add a test?

grasmash commented:
Yes, I’m working on a test and also trying to diagnose why the current tests are failing. But, I find that running tests while Xdebug is enableed is extremely slow with Symfony. Any tips on speeding that up?

grasmash commented:
So, the test is failing because it expects:
```
  0/50 [>---------------------------]   0%
\x1b[1A\x1b[0J  1/50 [>---------------------------]   2%
\x1b[1A\x1b[0J  2/50 [=>--------------------------]   4%
```

But the actual is:
```
  0/50 [>---------------------------]   0%
\x1b[2A\x1b[0J  1/50 [>---------------------------]   2%
\x1b[2A\x1b[0J  2/50 [=>--------------------------]   4%

```

I don't really understand the difference between the `\x1b[1A` and `\x1b[2A` strings.

grasmash commented:
Ok, so I did so research. This is a great article: https://notes.burke.libbey.me/ansi-escape-codes/. The 2A indicates that 2 lines should be cleared rather than 1. That seems right looking at the 3 line output in the example.

The rather unfun part of testing this is that you can't actually _see_ the diff for the failing tests on the command line because they're rendered ANSI escape codes and not actually visible. Apart from using xDebug to inspect the variables (which is terribly slow) I'm not sure how one is supposed to know what's being printed.

grasmash commented on src/Symfony/Component/Console/Helper/ProgressBar.php:
```diff
@@ -462,8 +462,14 @@ private function overwrite(string $message): void
+                    $message_lines = explode("\n", $message);
+                    $line_count = \count($message_lines) - $this->formatLineCount + 1;
+                    foreach ($message_lines as $message_line) {
+                        if (Helper::strlen($message_line) / $this->terminal->getWidth() > 1) {
```
This is a second bug that's being fixed. If the Terminal width exactly matches the line length, we should not count an extra line. We should only count an extra line if it exceeds the line length.

grasmash commented:
I don't believe new tests are needed. The existing test fixtures have been updated to assert that lines are emitted correctly. 

Admittedly just _looking_ at the test fixtures doesn't grant much confidence. I performed a number of successful manual tests that used combinations of short content, long content, and content with new line characters. They all displayed as expected. 

However, I'd like a functional review by someone else who is able to more readily identify issues.

grasmash commented on src/Symfony/Component/Console/Tests/Helper/ProgressBarTest.php:
```diff
@@ -337,8 +337,8 @@ public function testOverwriteWithSectionOutput()
             '  0/50 [>---------------------------]   0%'.\PHP_EOL.
-            "\x1b[1A\x1b[0J".'  1/50 [>---------------------------]   2%'.\PHP_EOL.
-            "\x1b[1A\x1b[0J".'  2/50 [=>--------------------------]   4%'.\PHP_EOL,
+            "\x1b[2A\x1b[0J".'  1/50 [>---------------------------]   2%'.\PHP_EOL.
```
This makes sense. 2 lines are cleared each time: 1 for the progress bar and 1 for the extra new line on the end.

grasmash commented on src/Symfony/Component/Console/Tests/Helper/ProgressBarTest.php:
```diff
@@ -364,9 +364,9 @@ public function testOverwriteMultipleProgressBarsWithSectionOutputs()
         $this->assertEquals(
             '  0/50 [>---------------------------]   0%'.\PHP_EOL.
             '  0/50 [>---------------------------]   0%'.\PHP_EOL.
-            "\x1b[1A\x1b[0J".'  1/50 [>---------------------------]   2%'.\PHP_EOL.
```
Again, the new test appears to make sense. 4 lines precede the line containing the clear character. The 4 lines are comprised of 2 PHP_EOLs and 2 strings.

grasmash commented:
@chalasr I believe this is ready to be merged. The failing tests seem unrelated to the changes made here.

chalasr commented:
Thank you for the detailed explanations, it makes everyone more comfortable with reviewing/merging such a patch (low-level + changes in existing tests).
 
Still, I'd like to checkout your branch to ensure we don't have any regression on the modified test cases (as it happened a lot in the past). I will look into this by the end of the week.

grasmash commented:
> Still, I'd like to checkout your branch to ensure we don't have any regression on the modified test cases (as it happened a lot in the past). I will look into this by the end of the week.

Completely agree. 👍

grasmash commented:
Separately, it looks like I should merge my approach with this proposed change: https://github.com/symfony/symfony/pull/40450/files

Done :)

chalasr commented:
Thanks, I was about to ask you if you could have a look at #40450 :)
Also can you rebase your PR on the 4.4 branch (+ change the base branch by editing your PR)?

grasmash commented:
I can rebase it, but I would prefer to have that in a separate pull request. Would that be OK?

grasmash commented:
Need to do some more work here to correctly integrate changes from #40450, seems to have caused a problem.

grasmash commented:
@chalasr Ok. Lot's of changes, but much better now:

1. I've added a new test and left existing fixtures as-is.
2. I've rebased on to 4.4 and changed the base branch for the PR.
3. I've cherry-picked @danepowell's new test from #40450.

grasmash commented:
I think that all test failures are unrelated to the changes. You can see the relevant tests passing here:
https://travis-ci.com/github/symfony/symfony/jobs/491226919#L8401

danepowell commented:
As the owner of #40450, I think this approach makes sense.

danepowell commented:
I guess I should say, I'd prefer for #40450 to be merged first since it addresses a separate bug and I'd love to get credit for the fix 😄 . But as long the bugs get fixed (this PR now fixes both), that's what matters most.

grasmash commented:
@danepowell I'm fine with that! Though I should say, I intentionally cherry picked your commit so you'd get commit credit if this PR is merged as is.

nicolas-grekas commented:
I merged #40450 to ease with squashing this PR.

grasmash commented:
Rebased.

chalasr commented on src/Symfony/Component/Console/Helper/ProgressBar.php:
```diff
@@ -441,8 +441,15 @@ private function overwrite(string $message): void
                 if ($this->output instanceof ConsoleSectionOutput) {
-                    $lines = floor(Helper::strlenWithoutDecoration($this->output->getFormatter(), $message) / $this->terminal->getWidth()) + $this->formatLineCount + 1;
-                    $this->output->clear($lines);
+                    $message_lines = explode("\n", $message);
```
Symfony exclusively uses camel case for variable names

chalasr commented:
@grasmash Can you squash your 7 commits into 1 so that we can merge?

grasmash committed:
Correctly clear lines for multi-line progress bar messages.

grasmash commented:
@chalasr Done!

fabpot commented:
Thank you @grasmash.
